### PR TITLE
ISolver interface and factory implemented

### PR DESCRIPTION
**Pull request by AlexanderSergeev**

### CODE REVIEW

File: Src/Optimization/ISolver.h [hunk 1]
```diff
@@ -3,9 +3,17 @@
...
 #include "IFunction.h"
+#include "Test.h"
 
 class ISolver
 {
 public:
-    virtual double solve(const IFunction& f, double a, double b, double epsilon, int n) = 0;
+    virtual void performTest(IFunction& f, double a, double b, double epsilon, int n) = 0;
+    virtual void evaluateSolution() = 0;
+    virtual bool checkStopCriterion(double epsilon, int n) = 0;
```
Comment (Zebraside) [inline Src/Optimization/ISolver.h]: IMO this is protected methods. Because users should know about them

File: Src/Optimization/ISolver.h [hunk 2]
```diff
@@ -3,9 +3,17 @@
...
+    virtual double solve(IFunction& f, double a, double b, double epsilon, int n, std::vector<Test>& history, int& operationCount) = 0;
```
Comment (Zebraside) [inline Src/Optimization/ISolver.h]: `std::vector<Test>& history, int& operationCount` what as a user should I pass  here? Why this variables in the public interfaec?

File: Src/Optimization/PiyavskogoMethod.cpp [hunk 3]
```diff
@@ -10,7 +10,7 @@ PiyavskogoMethod::PiyavskogoMethod(double _r)
 }
 
 
-double PiyavskogoMethod::solve(Function& f, double a, double b, double epsilon, int n, std::vector<Test>& history, int& operationCount)
+double PiyavskogoMethod::solve(IFunction& f, double a, double b, double epsilon, int n, std::vector<Test>& history, int& operationCount)
```
Comment (Zebraside) [inline Src/Optimization/PiyavskogoMethod.cpp]: Not there is no functions that can be passed here in the whole project. Could you also inherit Function from IFunction?

**Review by Zebraside — CHANGES_REQUESTED** [review]

File: Src/Optimization/PiyavskogoMethod.cpp (hunk 3 above)
Comment (AlexanderSergeev) [reply Src/Optimization/PiyavskogoMethod.cpp]: @Zebraside, fixed the issues. Please, review.

Comment (AlexanderSergeev) [reply Src/Optimization/PiyavskogoMethod.cpp]: @Zebraside, fixed the comments. Please, review.

**Review by Zebraside — APPROVED** [review]

**Pull request merged**